dlib/interfaces/map_pair.h

```diff
@@ -31,6 +31,9 @@ namespace dlib
         typedef T1 key_type;
         typedef T2 value_type;
 
+        map_pair() = default;
+        map_pair(const map_pair&) = default;
```

Comment:
Not sure if the assignment operator should be `=delete`d? (Now it's made unavailable by making it `protected`.)

Reply:
I mean this one here: https://github.com/davisking/dlib/pull/2490#pullrequestreview-872772748

Reply:
Yeah if it compiles with `=deleted` that's fine.  I forget why it's `protected` instead of `private`.  I wager mightily it's something like "because in some version of visual studio from the late 1990s it wouldn't build otherwise" since this is really old code.  I think I wrote it in 2001 and would have been running some version of visual studio on windows 98 probably.  Those were dark times in compilers :(

Reply:
`map_pair(const map_pair&) = delete` since this should be non-copyable.

Reply:
Indeed. No reason from my side to `=delete` the assignment operator, though.

Reply:
> `map_pair(const map_pair&) = delete` since this should be non-copyable.

I'm afraid that in that case some code (inside the dlib repo) didn't compile.

Perhaps we need to explicitly make this movable?

Reply:
Yeah, we can't just `=delete` it. If we do, we get these kinds of errors:
https://github.com/davisking/dlib/runs/5077995836

Reply:
Er, yeah. I just looked at this more and the assignment operator should remain protected as it is.  The defaulted constructors you added are fine too.